# Meeting notes 05-10-2020

Look up **Hardy-Weinberg Equilibrium**


# 19-10-2020

BRCA1
BRCA2

- model individually 

What is biologically plausible ?

- Use HG-38 reference genome 
- use frequency of each variant 
- use edit distance  
- distribution of variance depending on population 
- think about ethnicity data (data should inform the discriminator) (population frequencies) 
- pathogenicity of a variant BRCAexchange.org
    - `clinical_significance_ENAGMA` benign, likely benign, pathogenic, unknown 
- mis-sense vs non-sense vs synonymous variants 
    - *synonymous*: mutation that creates the same amino acid 
    - *mis-sense*: create different amino acide
    - *non-sense*" early STOP codon
- location of mutations:
    - functional domain or not
- cohort frequency: compare to frequency in the specific dataset


# 16-11-2020


- Try with just SNPs (easy to code)
    - makes model evaluation easier 

- Try FASTARefMaker, if does not work just modify reference with SNPs from files

- It's biologically plausible to model each of a person's haplotypes independantly
    - just draw twice from the generator for each person

# 12-1-2020

We treat the haplotypes as independant.


Hyperparameter tuning for the GAN:
- tune using prior knowledge of expected parameters

# 7-12-2020

Model evaluation:

- Does it preserve intron/exon structure with splice sites?
    - MaxEntScan to detect number of splice sites
        - needs GATK to get data back in variant calling format
    - are functional domains conserved (just compare regions for equality)
        - also should be in variant calling format
    - edit distance 
    - are both variants dysfunctional (future work!)
    - ethnicity (is the variant similar to that from the population from that ethnicity) (future work!)

How many unique genes does the network generate?

The code can incorporate simple subsititions easily (1 -> n or n->1)

Other variant calling tool: [hail](https://hail.is/docs/0.2/genetics/hail.genetics.Call.html)
- use relatedness tool in hail [link](https://hail.is/docs/0.2/methods/relatedness.html)


## 15-12-2020 (end or quarter presentations)

- Only look at functional domains that display variation in the training data


### Evaluating differential privacy performance 

Can a discriminator be trained to recognize a sequence generated by simply
taking variant sample frequencies
(c. Benedict Paten)


## 11-01-2021

**Cohort**: group in a particular study
 
**Population**: sum of all cohorts 


## 20-01-2021

Is is possible to characterise how much data is needed to get a "good"
generative model?


# 25-01-2021

## *Co-occurence analysis* software:

https://github.com/BRCAChallenge/federated-analysis

- ignore pathology report

- nohup [command] &

Look for concurrence with James' results


# 03-02-2021

- train on protein space
- train just on exome


- look at discriminator first
- train discriminator on output of a "simpler" generative model to evaluate model capacity 

- maybe autoencoder (simpler training than GAN?)
    - treat hidden dimension as a hyperparameter

- RNN based autoencoder
    - look at sequence autoencoders based on RNNs 
    - find minimum processing chunk that a CNN based architecture can take, and 
    - Batch norm before activation in ResNet
    

    - Look at where GANs are zeroed!
    - estimating information complexity of genomics sequences
    - 

# 06-05-2021

- Instead of conditioning one variant on the other, use information on whether variants 
are benign or pathogenic. Filter out all samples that are (doubly) known pathogenic as a post-processing 
step 


