# Notes on Papers

## Creating Artificial Human Genomes Using GenerativeModels

Almost never done in population genetics

What are selection tests?

**Single nucleotide polymorphisms (SNP)**: most common type of genetic variation, difference of a signle nucleotide

**Uniform Manifold Projections (UMAP):**: cf wikipedia


GANs perform poorly for low frequency alleles, but beats RBM overall on distance to real genome
GANs had excess of short chunks

**Haplotype:** group of genes within an organism that was inherited together from a single parent (just mom, or just dad)

**Genotype:** mom + dad

**SNP Distance**: # of different nucleotides

**Linkage Desiquilibrium:**


Metrics for of resemblance and privacy:
- **nearest neighbour adverserial accuracy ($AA_{TS}$):** measures whether two datasets were 191 generated by the same distribution based on the distances between all data points and 192 their nearest neighbours in each set
    - 0-0.5: overfitting, privacy loss 
    - 0.5-1: underfitting
    - 0.5: indistinguishable
TODO add mathematical detail

GANs underfit slightly
RBMS overfit slightly

Computational limitations prevent from creating full genomes with either method
However convolutional GANS can be parallelized w/o disrupting haplotype structure


- **privacy loss**:
TODO add mathematical detail


## Generating and designing DNA with deep generative models

Used a WGAN that inspired from a natural language model.
Code for original WGAN https://github.com/igul222/improved_ wgan_training

Using approach on DNA sequences requires care in handling discretozation:
- use continuous relaxation of one-hot sequence vectors with normalizad
probability distribution over nucleotides
- Recover a discrete sequence using argmax

Use edit distance to evaluate model perfomance
Reflecting latent space $z \rightarrow -z$ shows that model learn 
$A-T$ and $C-G$ bindings 

Can use **Position Weight Matrices** (PWM) to encode specific DNA motifs
into the predictor function
- 
- make using population allele frequency!


Claim that VAE is unsuitable to sequences, tested VAEs in multiple configurations
- VAE performs poorly in noisy scenarios

Paper code: https://github.com/co9olguy/Generating-and-designing-DNA


## Differentially private genetic models


## VCF Data format

Has reference genome, then list of positions differing from reference

Standard parsers: `pyVCF`, `scikit-allel`


## Privacy-preserving techniques of genomic data a survey

DP-AuGM is effective at defending against GAN attacks

Use MNIST dataset and hospital "tabular" data, but there is evidence that these
models do not perform well for discrete sequences --> maybe approach in that paper 
is not that useful



# How to generate DNA 19-10-2020

BRCA1
BRCA2

- model individually 

What is biologically plausible ?

- Use HG-38 reference genome 
- use frequency of each variant 
- use edit distance  
- distribution of variance depending on population 
- think about ethnicity data (data should inform the discriminator) (population frequencies) 
- pathogenicity of a variant BRCAexchange.org
    - `clinical_significance_ENAGMA` benign, likely benign, pathogenic, unknown 
- mis-sense vs non-sense vs synonymous variants 
    - *synonymous*: mutation that creates the same amino acid 
    - *mis-sense*: create different amino acide
    - *non-sense*" early STOP codon
- location of mutations:
    - functional domain or not
- cohort frequency: compare to frequency in the specific dataset
